libs/community/langchain_community/chat_models/notdiamond.py

```diff
@@ -0,0 +1,416 @@
+    elif isinstance(message, SystemMessage):
+        message_dict = {"role": "system", "content": message.content}
+    elif isinstance(message, FunctionMessage):
+        message_dict = {
+            "role": "function",
+            "content": message.content,
+            "name": message.name,
+        }
+    else:
+        raise ValueError(f"Got unknown type {message}")
+    if "name" in message.additional_kwargs:
+        message_dict["name"] = message.additional_kwargs["name"]
+    return message_dict
+
+def _convert_delta_to_message_chunk(
+    _dict: Mapping[str, Any], default_class: Type[BaseMessageChunk]
+) -> BaseMessageChunk:
+    role = _dict.get("role")
+    content = _dict.get("content") or ""
+    if _dict.get("function_call"):
+        additional_kwargs = {"function_call": dict(_dict["function_call"])}
+    else:
+        additional_kwargs = {}
+
+    if role == "user" or default_class == HumanMessageChunk:
+        return HumanMessageChunk(content=content)
+    elif role == "assistant" or default_class == AIMessageChunk:
+        return AIMessageChunk(content=content, additional_kwargs=additional_kwargs)
+    elif role == "system" or default_class == SystemMessageChunk:
+        return SystemMessageChunk(content=content)
+    elif role == "function" or default_class == FunctionMessageChunk:
+        return FunctionMessageChunk(content=content, name=_dict["name"])
+    elif role or default_class == ChatMessageChunk:
+        return ChatMessageChunk(content=content, role=role)  # type: ignore[arg-type]
+    else:
+        return default_class(content=content)  # type: ignore[call-arg]
+    
+def completion_with_retry(
+    llm: ChatNotDiamond,
+    run_manager: Optional[CallbackManagerForLLMRun] = None,
+    **kwargs: Any,
+) -> Any:
+    """Use tenacity to retry the completion call."""
+    retry_decorator = _create_retry_decorator(llm, run_manager=run_manager)
+
+    @retry_decorator
+    def _completion_with_retry(**kwargs: Any) -> Any:
+        if "stream" in kwargs:
+            return llm.client.chat.completions.stream(**kwargs)
+        return llm.client.chat.completions.create(**kwargs)
+
+    return _completion_with_retry(**kwargs)
+
+async def acompletion_with_retry(
+    llm: ChatNotDiamond,
+    run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+    **kwargs: Any,
+) -> Any:
+    """Use tenacity to retry the async completion call."""
+    retry_decorator = _create_retry_decorator(llm, run_manager=run_manager)
+
+    @retry_decorator
+    async def _completion_with_retry(**kwargs: Any) -> Any:
+        return await llm.client.chat.completions.acreate(**kwargs)
+
+    return await _completion_with_retry(**kwargs)
+
+class ChatNotDiamond(BaseChatModel):
+    """Chat model that uses the Not Diamond SDK."""
+
+    client: Any = Field(default=None, exclude=True)  #: :meta private:
+    # Not Diamond parameters
+    llm_configs: List[str] = Field(min_items=1)
```

Comment:
Making this somewhat more concrete: in a `@pre_init` method, you could explore attaching `vars(client)` to the calling class itself.

```python

class ChatNotDiamond(BaseChatModel):

    @pre_init
    def validate_environment(cls, values: Dict) -> Dict:
        ...
        client = NotDiamond(...)
        for attr, attr_val in vars(client).items():
            setattr(cls, attr, attr_val)
```

Then, after the @pre_init runs:

```python
chat_nd = ChatNotDiamond(...)
chat_nd.latency_tracking  # returns default latency_tracking config from NotDiamond()
```

This could save us lots of maintenance, although I don't know if langchain's usage of pydantic.v1 will limit our ability to dynamically add attributes in this way. So we can skip it.